libyul/optimiser/Semantics.h

```diff
@@ -53,11 +53,49 @@ class SideEffectsCollector: public ASTWalker
 		Block const& _ast,
 		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
 	);
+	SideEffectsCollector(
```

Comment:
This class now has a bunch of different constructors for different AST nodes. Wouldn't it be better to use `operator()`s?

Reply:
Interesting. How will your proposal look like? We already have `void operator()(...)`

Reply:
Exactly, I'm suggesting to use those. But I guess the idea (even before this PR) was to have constructors kind of filter the intent and call `operator()` accordingly

Reply:
I think the constructor for `Expression` and `Statement` are useful because we would have to call visit otherwise. Maybe it would be better to remove the constructors for `Block` and `ForLoop`, though.